
# 🧰 Day 2: API Access and Python Requests Library Hands-on
In this Notebook, we going to explore
- What is API and its concepts
- What is python request library
- Basic HTTP requests-get requests with param, post requests with data, response status code
- Create .env file and add Openrouter API key to .env file
- Making First call to LLM using requests and understanding its response

## What is an API?
- An API, or Application Programming Interface, is a set of rules that defines how applications or devices can connect and communicate with each other.
APIs enable:

    * **Integration**: Allowing different applications to work together.
    * **Data Sharing**: Facilitating the transfer of data between systems.

### 📖 Understanding API 
| Concept            | Description / Purpose                                                                                                                                                                                                                                                                                                                                                           |
| :----------------- | :------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| **API URL (Base URL)** | The **foundational address** for an API service. It's the **common starting point** for all requests to that specific API.                                                                                                                                                                                                                                                                |
| **Endpoint** | A **specific path** appended to the API URL that represents a **particular resource** or **function** available through the API. It tells the server what specific action or data you are requesting. <br> **Example:** If Base URL is `https://api.example.com/v1`, an endpoint might be `/users` or `/products/123`.                                                                             |
| **Headers** | **Key-value pairs** sent in the request (and received in the response) that provide **metadata** about the **request or response**. They include things like **authentication tokens, content type, user agent information, and caching instructions**. <br> **Common Headers:** `Authorization`, `Content-Type`, `Accept`, `User-Agent`.                                                      |
| **Data (Body/Payload)** | The **main content or payload** sent in the request, typically used with HTTP methods like `POST`, `PUT`, `PATCH`. This contains the actual information you want to send to the server (e.g., new user details, data to be updated). <br> **Format:** Often JSON or XML.                                                                                                            |
| **Query Parameters** | **Key-value pairs appended to the URL after a `?`** (and separated by `&`) that provide **additional filtering, sorting, or pagination criteria** for the request. They are typically used with `GET` requests to modify the response. <br> **Example:** `https://api.example.com/v1/products?category=electronics&limit=10&page=2` (`category`, `limit`, `page` are query params). |
| **Response** | The **data sent back by the API server** in reply to your request. It includes: <br> - **Status Code:** An HTTP status code (e.g., 200 OK, 404 Not Found, 500 Internal Server Error) indicating the result of the request. <br> - **Headers:** Metadata about the response. <br> - **Body (Payload):** The actual data requested, typically in JSON or XML format.                    |

****

## Introduction to the Python Requests Library
- The requests library is a simple, yet **powerful, HTTP library for Python**. It allows you to **send HTTP/1.1 requests** easily.
- It offers intuitive **methods** like **get, post, put, delete** corresponding to HTTP verbs. 
- You provide a URL and optional data/parameters, and it handles the underlying HTTP complexities.
- The **response** from the server is **returned as a Response object**, containing **status codes, headers, and the content** (text, bytes, or JSON). 
- It's a user-friendly way to interact with web services and APIs in Python.

## Basic HTTP requests with GET, POST methods, Response and status code
Lets install python request, make a simple python request call and explore its components
##### Note: Please select kernel/virtual environment, before executing the code

In [39]:
! pip install -q requests


[notice] A new release of pip available: 22.3.1 -> 25.1.1
[notice] To update, run: python.exe -m pip install --upgrade pip


#### Import dependencies

In [1]:
import requests

#### GET request
- Purpose: Retrieves data from a server.

In [5]:
# Base URL and endpoint 
url = "https://jsonplaceholder.typicode.com/posts"


# Make GET request with params
response = requests.get(url)

# Print status code and JSON response
print("Status Code:", response.status_code)
print('-'*20)
print("URL with Params:", response.url)
print('-'*20)


Status Code: 200
--------------------
URL with Params: https://jsonplaceholder.typicode.com/posts
--------------------


In [9]:
# Print the JSON response
response.json()

[{'userId': 1,
  'id': 1,
  'title': 'sunt aut facere repellat provident occaecati excepturi optio reprehenderit',
  'body': 'quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto'},
 {'userId': 1,
  'id': 2,
  'title': 'qui est esse',
  'body': 'est rerum tempore vitae\nsequi sint nihil reprehenderit dolor beatae ea dolores neque\nfugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis\nqui aperiam non debitis possimus qui neque nisi nulla'},
 {'userId': 1,
  'id': 3,
  'title': 'ea molestias quasi exercitationem repellat qui ipsa sit aut',
  'body': 'et iusto sed quo iure\nvoluptatem occaecati omnis eligendi aut ad\nvoluptatem doloribus vel accusantium quis pariatur\nmolestiae porro eius odio et labore et velit aut'},
 {'userId': 1,
  'id': 4,
  'title': 'eum et est occaecati',
  'body': 'ullam et saepe reiciendis voluptatem adipisci\nsit amet autem assumenda provid

#### GET request with params
- Purpose: Retrieves data from a server with filter.

In [10]:
# Base URL and endpoint 
url = "https://jsonplaceholder.typicode.com/posts"

# Query parameters
params = {
    'userId': 1
    }

# Headers
headers = {'content-Type':'application/json'}

# Make GET request with params
response = requests.get(url, params=params, headers = headers)

# Print status code and JSON response
print("Status Code:", response.status_code)
print('-'*20)
print("URL with Params:", response.url)
print('-'*20)

Status Code: 200
--------------------
URL with Params: https://jsonplaceholder.typicode.com/posts?userId=1
--------------------


In [12]:
# Print the JSON response
response.json()

[{'userId': 1,
  'id': 1,
  'title': 'sunt aut facere repellat provident occaecati excepturi optio reprehenderit',
  'body': 'quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto'},
 {'userId': 1,
  'id': 2,
  'title': 'qui est esse',
  'body': 'est rerum tempore vitae\nsequi sint nihil reprehenderit dolor beatae ea dolores neque\nfugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis\nqui aperiam non debitis possimus qui neque nisi nulla'},
 {'userId': 1,
  'id': 3,
  'title': 'ea molestias quasi exercitationem repellat qui ipsa sit aut',
  'body': 'et iusto sed quo iure\nvoluptatem occaecati omnis eligendi aut ad\nvoluptatem doloribus vel accusantium quis pariatur\nmolestiae porro eius odio et labore et velit aut'},
 {'userId': 1,
  'id': 4,
  'title': 'eum et est occaecati',
  'body': 'ullam et saepe reiciendis voluptatem adipisci\nsit amet autem assumenda provid

#### POST method request with data
- Purpose: Submits data to a server to create a new resource or perform an action.

In [42]:
import requests

# Base URL and endpoint for creating a post
url = "https://jsonplaceholder.typicode.com/posts"

# Headers
headers = {
    "Content-Type": "application/json",
    "Authorization": "Bearer YOUR_TOKEN_HERE"
}

# JSON data to send 
data = {
    "title": "Foo",
    "body": "Bar content here",
    "userId": 1
}

# Make POST request
response = requests.post(url, json=data, headers=headers)

# Print status code and returned data
print("Status Code:", response.status_code)
print('-'*50)
print("Response JSON:")
print(response.json())

Status Code: 201
--------------------------------------------------
Response JSON:
{'title': 'Foo', 'body': 'Bar content here', 'userId': 1, 'id': 101}


### Difference between GET requests with params, POST requests with json/data
| Feature                | GET with `params`                    | POST with `json`          |
| ---------------------- | ------------------------------------ | ------------------------- |
| Data location          | URL query string                     | Request body              |
| Primary purpose        | Retrieve data                        | Create/modify data        |
| Caching                | Can be cached                        | Not cached                |
| Data length            | Limited by URL length (\~2000 chars) | No strict limit (in body) |
| Secure for credentials | ❌ No                                 | ✅ Yes (when over HTTPS)   |
| Visibility             | Data visible in URL                  | Hidden from URL (in body) |

### Different request methods
| Method | Definition                                                                 | Common Use Cases                         |
|--------|----------------------------------------------------------------------------|-----------------------------------------|
| `GET`  | Retrieve data from a specified resource.                                    | Fetching web pages, APIs, files.        |
| `POST` | Submit data to be processed (e.g., form data, files).                       | Submitting forms, creating resources.   |
| `PUT`  | Replace the entire resource at the specified URI.                           | Updating resources (full replacement).  |
| `DELETE`| Remove the specified resource.                                              | Deleting records/files.                 |
| `HEAD` | Same as `GET`, but retrieves only headers (no response body).               | Checking resource metadata/headers.     |
| `PATCH`| Partially update a resource (only modified fields).                         | Partial updates (e.g., API fields).     |
| `OPTIONS`| Describe communication options for the target resource (CORS configuration).| Checking supported methods/headers.     |


### Headers keys and its purpose
| Header          | Purpose                                                                                                         |
| --------------- | --------------------------------------------------------------------------------------------------------------- |
| `Content-Type`  | Specifies the format of the data you're sending (`application/json`, `application/x-www-form-urlencoded`, etc.) |
| `Accept`        | Tells the server what response format you want (`application/json`, `text/html`, etc.)                          |
| `Authorization` | Used for sending API keys, tokens (Bearer, Basic auth, etc.)                                                    |
| `User-Agent`    | Identifies the client (browser, app, script, etc.)                                                              |
| `Referer`       | Tells the server where the request originated from                                                              |

### Handling Responses
- response.status_code: HTTP status code (200 for OK, 404 for Not Found, etc.).
- response.json(): Parse JSON content.
- response.text: Get the response content as a string.
- response.headers: Access response headers.

### Response Status code
| Status Code | Category      | Meaning                                     | Common Use                                                               |
| :---------- | :------------ | :------------------------------------------ | :----------------------------------------------------------------------- |
| **1xx** | **Informational** | Request received, continuing process.       | **Rarely seen** by end-users.                                                |
| 100         | Informational | Continue                                    | The server has received the request headers and the client should proceed to send the request body. |
| **2xx** | **Success** | The action was successfully received, understood, and accepted. | Indicates a **successful operation**.                                        |
| 200         | Success       | OK                                          | Standard response for successful HTTP requests.                          |
| 201         | Success       | Created                                     | The request has been fulfilled and resulted in a new resource being created. |
| 204         | Success       | No Content                                  | The server successfully processed the request, but is not returning any content. |
| **3xx** | **Redirection** | Further action needs to be taken to complete the request. | Used for **URL redirects**.                                                  |
| 301         | Redirection   | Moved Permanently                             | The requested resource has been assigned a new permanent URI.            |
| 302         | Redirection   | Found (Temporary Redirect)                  | The requested resource is temporarily under a different URI.             |
| 304         | Redirection   | Not Modified                                | The resource has not been modified since the version specified by the request headers. (Caching). |
| **4xx** | **Client Error** | The request contains bad syntax or cannot be fulfilled. | Indicates an error originating from the client.                          |
| 400         | Client Error  | Bad Request                                 | The server cannot or will not process the request due to an apparent client error. |
| 401         | Client Error  | Unauthorized                                | Authentication is required and has failed or has not yet been provided. |
| 403         | Client Error  | Forbidden                                   | The server understood the request but refuses to authorize it. (e.g., insufficient permissions). |
| 404         | Client Error  | Not Found                                   | The requested resource could not be found on the server.                 |
| 405         | Client Error  | Method Not Allowed                          | The method specified in the request (e.g., GET, POST) is not allowed for the resource identified by the URI. |
| 429         | Client Error  | Too Many Requests                           | The user has sent too many requests in a given amount of time ("rate limiting"). |
| **5xx** | **Server Error** | The server failed to fulfill an apparently valid request. | Indicates an error on the server side.                                   |
| 500         | Server Error  | Internal Server Error                       | A generic error message, given when an unexpected condition was encountered and no more specific message is suitable. |
| 502         | Server Error  | Bad Gateway                                 | The server, while acting as a gateway or proxy, received an invalid response from an inbound server. |
| 503         | Server Error  | Service Unavailable                         | The server is currently unable to handle the request due to temporary overloading or maintenance. |
| 504         | Server Error  | Gateway Timeout                             | The server, while acting as a gateway or proxy, did not receive a timely response from an upstream server. |

***

## 🔐 Create .env file and add Openrouter API key to .env file

We’ll use `.env` files and the `python-dotenv` package to manage secrets like API keys.

#### What is .env File?
* The .env file is a plain text file used to store environment variables that are specific to a project.
* Inside the .env file, variables are defined as key-value pairs, one per line, typically in the format KEY=VALUE.

#### Why Use .env Files?
* Keep sensitive data out of source code
* Easy to manage different configurations for development/production
* Follows security best practices

#### Step 1: Create `.env` 

Open notepad or create a file in vscode and put below content and save it as `.env` file in the root of your project directory
Or add below content in newline if `.env` already available.
```
# OpenRouter API Configuration for OpenAI SDK
OPENROUTER_API_KEY=your_api_key_here

# Base URL for OpenRouter API
OPENROUTER_BASE_URL=https://openrouter.ai/api/v1

# Optional: Application identification for OpenRouter rankings
# HTTP_REFERER=https://your-site.com
# X_TITLE=Your App Name
```

This file should be added to `.gitignore`.

##### Note: Replace `your_api_key_here`with API key generated in Openrouter

#### Step 2: create `.gitignore` and dd `.env` to `.gitignore`

Open notepad or create a new file in vscode and put below content and save it as `.env` file in the root of your project directory.
Or add below content in newline if `.gitignore` already available.

```
.env
```

#### Step 3: Lets check we can access API key from `.env` file

##### Installing dependencies

In [43]:
! pip install -q python-dotenv


[notice] A new release of pip available: 22.3.1 -> 25.1.1
[notice] To update, run: python.exe -m pip install --upgrade pip


##### Import dependencies

In [15]:
from dotenv import load_dotenv
import os

In [16]:
load_dotenv()
api_key = os.getenv("OPENROUTER_API_KEY")
print("API Key Loaded:", bool(api_key))

API Key Loaded: True


Note: If API key loaded returns false, this mean
* API key not provided
* jupyter notebook not able to read the updated .env, restart the jupyter notebook 

****

## Making Your First LLM call using Python requests

- Now we call to LLM and check its response.


#### Step 1: Get API key, Openrouter url from `.env` file and assign it to variable

In [46]:
load_dotenv()

API_KEY = os.getenv("OPENROUTER_API_KEY")
OPENROUTER_BASE_URL = os.getenv("OPENROUTER_BASE_URL")

if API_KEY:
    print("✅ OpenAI API Key loaded from .env file.")
else:
    print("❌ OpenAI API Key not found in .env file.")

✅ OpenAI API Key loaded from .env file.


#### Step 2: Declare variable like URL, data, headers

In [ ]:
headers = {
    "Authorization": f"Bearer {API_KEY}",
    "Content-Type": "application/json"
}

# Simulated OpenAI API request

url = OPENROUTER_BASE_URL + "/chat/completions"  #OPENROUTER_BASE_URL=https://openrouter.ai/api/v1, endpoint ="/chat/completions"
print("API URL:", url)

data = {
    "model": "openai/gpt-3.5-turbo",
    "messages": [
        {"role": "user", "content": "what is the Tallest mountain on earth?"}
    ],
    "max_tokens": 50
}


API URL: https://openrouter.ai/api/v1/chat/completions


#### Step 3: Initiate the POST request call

In [21]:
try:
    response = requests.post(url, headers=headers, json=data)
    print("✅ Status:", response.status_code)
    print("🧠 Response:\n", response.json())
except Exception as e:
    print("❌ API call failed:", e)


✅ Status: 200
🧠 Response:
 {'id': 'gen-1749090716-vC52karQqp76ZNNK3NUS', 'provider': 'OpenAI', 'model': 'openai/gpt-3.5-turbo', 'object': 'chat.completion', 'created': 1749090716, 'choices': [{'logprobs': None, 'finish_reason': 'stop', 'native_finish_reason': 'stop', 'index': 0, 'message': {'role': 'assistant', 'content': 'Mount Everest, located in the Himalayas on the border of Nepal and China, is the tallest mountain on earth with a height of 29,031.7 feet (8,848.86 meters) above sea level.', 'refusal': None, 'reasoning': None}}], 'system_fingerprint': None, 'usage': {'prompt_tokens': 16, 'completion_tokens': 45, 'total_tokens': 61, 'prompt_tokens_details': {'cached_tokens': 0}, 'completion_tokens_details': {'reasoning_tokens': 0}}}


#### 📬 What Does a Response Look Like?

LLM APIs response structure:

```json
{
    'id': 'gen-1749002670-fs2k0cNyGYC2ABipRp4A', 
    'provider': 'OpenAI',
    'model': 'openai/gpt-3.5-turbo', 
    'object': 'chat.completion', 
    'created': 1749002670, 
    'choices': [
        {'logprobs': None, 
        'finish_reason': 'stop', 
        'native_finish_reason': 'stop',
        'index': 0, 
        'message': {
                    'role': 'assistant', 
                    'content': 'Mount Everest, located in the Himalayas on the border between Nepal and China, is the tallest mountain on Earth. It stands at a height of 29,032 feet (8,848 meters) above sea level.', 'refusal': None, 'reasoning': None
                   }
        }
        ], 
    'system_fingerprint': None, 
    'usage': {
                'prompt_tokens': 16,  //what is the Tallest mountain on earth?
                'completion_tokens': 44, 
                'total_tokens': 60, 
                'prompt_tokens_details': {'cached_tokens': 0}, 
                'completion_tokens_details': {
                                                'reasoning_tokens': 0
                                            }
            }
}



#### 🧩 Step 4: Parsing JSON Responses

Let’s extract specific values from the JSON.

In [23]:
response_json =  response.json()

answer = response_json["choices"][0]["message"]["content"]
print("User's Question: what is the Tallest mountain on earth?")
print("🧠 Assistant's Response:\n", answer)

User's Question: what is the Tallest mountain on earth?
🧠 Assistant's Response:
 Mount Everest, located in the Himalayas on the border of Nepal and China, is the tallest mountain on earth with a height of 29,031.7 feet (8,848.86 meters) above sea level.


### What is a Token?
- In LLMs, a token is the basic unit of text the model processes. 
- It's often a whole word, part of a word, or a punctuation mark.
- Models understand language by breaking text into these tokens. 
- Tokens are the fundamental currency of LLMs, determining both their processing capacity and the cost of using them.
- Pricing and context limits are based on token counts.

Example 1:
- The sentence **"Hello, world!"** might be tokenized as:
    * "Hello"
    * ","
    * " " (space)
    * "world"
    * "!"
- This would be 5 tokens.

Example 2:
- The word **"unbelievable"** might be tokenized as:
    * "un"
    * "believ"
    * "able"
- This would be 3 tokens.

***